design-proposals/vm-rollout-strategy/vm-rollout-strategy.md

```diff
@@ -0,0 +1,207 @@
+    
+
+## Non Goals
+
+-   Live uppdate for **all** fields of the Virtual Machine API. Only select fields can be propagated to the guest. 
+-   Make LiveUpdate the default. We defer this decision.
+-   Support upgrades of VMs that are using the `liveUpdateFeatures`.
+-   This design does consider the hot-plugging of network interfaces and additional storage.
+
+## Definition of Users
+
+A VM owner - with (at a minimum) create, update, and patch access for VM objects within a namespace.
+
+
+## User Stories
+
+As a VM owner I want changes that I applied to the VM object to propagate to the VMI if possible.
+
+As a VM owner, I want to be notified about changes that require a restart of the VM.
+
+## Repos
+
+kubevirt/kubevirt
+
+
+# Design
+
+> **Note:** Below we are only discussing changes to fields that support live updates.
+                  References to `all` fields only refers to fields that can be live-updated. 
+
+## KubeVirt CR
+
+A new field called `vmRolloutStrategy` will be introduced in the KubeVirt CR. This field will control the behavior of the VM objects. Two possible values will be available:
+
+1. `LiveUpdate` means all changes to the VM object will be eventually applied to the VMI.
+    
+2. `Stage` would mean the current familiar behavior remains - all changes to the VM object will be staged and applied after a restart.
+
+In both cases, a condition called `RestartRequired` will be added to the VM to reflect the need for a restart and the reason why.
+
+The default value of the `vmRolloutStrategy` field if not specified will be `Stage`.
+
+The `vmRolloutStrategy` field will be usable only if the `VMLiveUpdateFeatures` feature gate is enabled.
+
+
+## VM API
+
+The VM level `liveUpdateFeatures` field will be removed.
+VMs using the `liveUpdateFeatures` will be rejected by the API.
+
+If `RestartRequired` condition is present, any additional changes to the VM spec - to live-updatedable or non-liveupdateable fields - will be staged. Thus the VM has to be restarted.
+
+## Changes to CPU/Memory hotplug workflow
+
+### CPU Hot Plug
+Since the LiveUpdateFeatures API will be deprecated the `spec.template.spec.domain.cpu.maxSockets`field of the VM object will become mutable and stays optional.
+If it is not set by the user then the maxSockets value will continue to be internally computed according the cluster-wide `maxHotPlugRatio`, as it is done today.
+
+Changes to `spec.template.spec.domain.cpu.maxSockets` would require a restart of the VM. This would also set the `RestartRequired` condition.
+
+CPU hot plug is triggered by changing `spec.template.spec.domain.cpu.sockets`. This change will be propagated to the VMI object as long as its value is lower than the value of `MaxSockets` on the last VM startup.
+Increasing the value of the `spec.template.spec.domain.cpu.sockets` beyond the `MaxSockets` will require a restart to take effect.
+
```

Comment:
Long discussion sumary here:
In theory Sockets are not needed for hotplug - the # of vCPUs would do. As this has quite some implications, going with sockets seems to be fine, but that also means that in case of a topology like 32 Cores/2Threads only 64! additional CPUs can be added (or removed) - which is often not wanted.

Ideally we will get this to as granular as a single vCPU  (as qemu fills up the topology if there is not a complete socket popuated on its own).

Reply:
Yes, we would like to take this vCPU approach to a dedicated proposal.

Reply:
Qemu alone cannot play a role in determining the VM's topology as KubeVirt needs to manage it correctly to assign vCPUs to a cpuset it receives from k8s. 
This is required for VMs with dedicated CPUs of course. 
Also, the current VM API requires setting a topology and a numb of vCPUs.
KubeVirt can add a vCPUs count but it won't be done as part of this proposal.

Reply:
We'll follow up on the vCPU discussion seaparately